### PR TITLE
✨ Add Diff Functionality

### PR DESCRIPTION
**Pull request by SteffenKn**
## What did you change?

This PR adds diff functionality to compare the current diagram with the last saved diagram.

Also fixes the bug that the detail view did not get detached and attached when changing the process definition via the solution explorer.

Relates to:

https://github.com/process-engine-pro/organization/issues/154

## How can others test the changes?

- Open up the BPMN-Studio
- Go to the Detail View of any diagram
- Add, remove, move and change some elements
- Go to the Diff View
- Have a look at the diff and the two diff modes

## Features that should be implemented later:

- Changeable diff colors
- Moving to Element when clicked on change in change list
- Resize functionality for the Viewers

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [ ] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  public savedxmlChanged(): void {
+    this._importXml(this.savedxml, this._leftViewer);
+  }
+
+  public changesChanged(): void {
+    this._updateDiffView();
+    this._prepareChangesForChangeList();
+  }
+
+  private _syncAllViewer(): void {
```
Comment (Paulomart) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: `_syncAllViewers()`

File: src/modules/bpmn-diff-view/bpmn-diff-view.html [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        <i class="fas fa-arrow-down fa-3x"></i>
+      </span>
+      <div class="diff-change-list" if.bind="showChangeList">
+        <span class="no-change-message" if.bind="noChangesExisting">${noChangesReason}</span>
+        <div class="change-list-table-wrapper">
+          <table class="change-list-table">
+            <tr repeat.for="removedElement of changeListData.removed">
+              <td class="change-list-table__entry">
+                <span class="label label-red">Removed</span> - ${removedElement.type}
+                <span if.bind="removedElement.name !== ''">mit dem Namen ${removedElement.name}
```
Comment (LeoTT) [inline src/modules/bpmn-diff-view/bpmn-diff-view.html]: illegal html, span does not end.

File: src/modules/bpmn-diff-view/bpmn-diff-view.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+              </td>
+            </tr>
+            <tr repeat.for="changedElement of changeListData.changed">
+              <td class="change-list-table__entry">
+                <span class="label label-orange">Change</span> - ${changedElement.type}
+                <span if.bind="changedElement.name !== ''">mit dem Namen ${changedElement.name}
```
Comment (LeoTT) [inline src/modules/bpmn-diff-view/bpmn-diff-view.html]: see above

File: src/modules/bpmn-diff-view/bpmn-diff-view.html [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+              </td>
+            </tr>
+            <tr repeat.for="addedElement of changeListData.added">
+              <td class="change-list-table__entry">
+                <span class="label label-green">Added</span> - ${addedElement.type}
+                <span if.bind="addedElement.name !== ''">mit dem Namen ${addedElement.name}
```
Comment (LeoTT) [inline src/modules/bpmn-diff-view/bpmn-diff-view.html]: see above

File: src/modules/bpmn-diff-view/bpmn-diff-view.html [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+              </td>
+            </tr>
+            <tr repeat.for="layoutChangedElement of changeListData.layoutChanged">
+              <td class="change-list-table__entry">
+                <span class="label label-purple">Layout Changed</span> - ${layoutChangedElement.type}
+                <span if.bind="layoutChangedElement.name !== ''">mit dem Namen ${layoutChangedElement.name}
```
Comment (LeoTT) [inline src/modules/bpmn-diff-view/bpmn-diff-view.html]: see above

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+  public showChangeList: boolean;
+  public noChangesExisting: boolean = true;
+  public noChangesReason: string;
+  public totalChangeAmount: number;
+  public changeListData: IDiffChangeListData = {
+    removed: [],
+    changed: [],
+    added: [],
+    layoutChanged: [],
+  };
```
Comment (LeoTT) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: public should be declared before private

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+  @bindable() public changes: IDiffChanges;
+  public leftCanvasModel: HTMLElement;
+  public rightCanvasModel: HTMLElement;
+  public lowerCanvasModel: HTMLElement;
+  public currentDiffMode: DiffMode;
+  public diffModeTitle: string = '';
+  public showChangeList: boolean;
+  public noChangesExisting: boolean = true;
+  public noChangesReason: string;
+  public totalChangeAmount: number;
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Maybe `totalAmountOfChanges` or `totalNumberOfChanges` sounds a bit better.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+    if (this.noChangesExisting) {
+      this._setNoChangesReason();
+    } else {
+      this.noChangesReason = '';
+    }
+  }
+
+  private _setNoChangesReason(): void {
+      const diagramsAreSame: boolean = this.savedxml === this.xml;
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: `diagramsAreTheSame`

File: src/contracts/bpmnmodeler/ICanvas.ts [hunk 9]
```diff
@@ -1,5 +1,7 @@
 import {IShape} from './IShape';
 
 export interface ICanvas {
```
Comment (Vyperus) [inline src/contracts/bpmnmodeler/ICanvas.ts]: Please add explaining comments to this interface.
See https://github.com/process-engine/solutionexplorer.repository.contracts/blob/develop/src/ISolutionExplorerRepository.ts

File: src/contracts/bpmnmodeler/IEventFunction.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+export interface IEventFunction {
```
Comment (Vyperus) [inline src/contracts/bpmnmodeler/IEventFunction.ts]: Please add explaining comments to this interface.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+  private _setEventFunctions(changingCanvas: ICanvas, firstCanvas: ICanvas, secondCanvas: ICanvas): void {
+    const changingCanvasContainer: HTMLElement = changingCanvas._container;
+
+    const adjustViewboxes: IEventFunction = (): void => {
+      const changedViewbox: string = changingCanvas.viewbox();
+      firstCanvas.viewbox(changedViewbox);
+      secondCanvas.viewbox(changedViewbox);
+    };
+
+    const checkForMousemovement: IEventFunction = (): void => {
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: typo: `checkForMouseMovement`

File: src/contracts/diffview/IDiffChanges.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export interface IDiffChanges {
```
Comment (Vyperus) [inline src/contracts/diffview/IDiffChanges.ts]: Please add explaining comments to this interface.

File: src/contracts/diffview/IDiffChangeListData.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import {IChangeListEntry} from './IChangeListEntry';
+
+export interface IDiffChangeListData {
```
Comment (Vyperus) [inline src/contracts/diffview/IDiffChangeListData.ts]: Please add explaining comments to this interface.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+      window.onmousemove = adjustViewboxes;
+    };
+    const stopCheckingForMousemovement: IEventFunction = (): void => {
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: If we have a `stop...` method for this, the `checkForMousemovement` Method should renamed to something like `startCheckingForMouseMovement`.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+        this.currentDiffMode = diffMode;
+        this._updateDiffView();
+      }),
+      this._eventAggregator.subscribe(environment.events.diffView.toggleChangeList, () => {
+        this.showChangeList = !this.showChangeList;
+      }),
+    ];
+  }
+
+  public created(): void {
```
Comment (Vyperus) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Please place the created() above the attached().

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+        changeListEntry = this._createChangeListEntry(elementChange.name, elementChange.$type);
+      }
+
+      changeListEntries.push(changeListEntry);
+    }
+
+    return changeListEntries;
+  }
+
+  private _prepareChangesForChangeList(): void {
```
Comment (Vyperus) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Some explaining comments would be nice.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+    } else if (this.currentDiffMode === DiffMode.BeforeVsAfter) {
+      this._updateLowerDiff(this.savedxml);
+      this.diffModeTitle = 'Before vs. After';
+    } else {
+      this.diffModeTitle = '';
+    }
+  }
+
+  private async _updateLowerDiff(xml: string): Promise<void> {
+    const xmlIsNotLoaded: boolean = xml === undefined || xml === null;
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: I would use at least one pair of parenthesis. Like so:
```
const xmlIsNotLoaded: boolean = (xml === undefined || xml === null)
```

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+    const diffModeIsAfterVsBefore: boolean = this.currentDiffMode === DiffMode.AfterVsBefore;
+
+    await this._importXml(xml, this._diffModeler);
+    this._markLayoutChangedElements(layoutChangedElements);
+    this._markChangedElements(changedElements);
+
+    diffModeIsAfterVsBefore ?
+      this._markAddedElements(addedElements) :
+      this._markRemovedElements(removedElements);
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: I would use am `if else` construct here:
```
 if (diffModeIsAfterVsBefore) {
    this._markAddedElements(addedElement);
} else {
    this._markRemovedElements(removedElements);
}
```

Ternary expressions are great when assigning stuff. But for a general control flow, I would always use `if else`.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+
+    const coloredXml: string = await this._getXmlFromModeler();
+    await this._importXml(coloredXml, this._lowerViewer);
+  }
+
+  private _importXml(xml: string, viewer: IBpmnModeler): Promise <void> {
+    const xmlIsNotLoaded: boolean = xml === undefined || xml === null;
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Like above, I would use one pair of parentheses here.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+
+    diffModeIsAfterVsBefore ?
+      this._markAddedElements(addedElements) :
+      this._markRemovedElements(removedElements);
+
+    const coloredXml: string = await this._getXmlFromModeler();
+    await this._importXml(coloredXml, this._lowerViewer);
+  }
+
+  private _importXml(xml: string, viewer: IBpmnModeler): Promise <void> {
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: I would define a method, that returns a promise, as `async` (event if the method only _returns_ a `void` Promise)

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+    const xmlIsNotLoaded: boolean = xml === undefined || xml === null;
+
+    if (xmlIsNotLoaded) {
+      return;
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Is this case a _normal_ case or can we throw an error here?

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+          reject(importXmlError);
+          return;
+        }
+
+        resolve();
+      });
+    });
+  }
+
+  private _getXmlFromModeler(): Promise<string> {
```
Comment (robpal94) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Like above, we should define this method as `async`.

File: src/modules/elementname/elementname.service.ts [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+  public getHumanReadableType(elementType: string): string {
+
+    const elementTypeWithoutBPMN: string = elementType.replace('bpmn:', '');
+    const humanReadableElementType: string = elementTypeWithoutBPMN.replace(/([A-Z])/g, ' $1').trim();
+
+    return humanReadableElementType;
+  }
+
+  public getHumanReadableName(elementName: string): string {
+    const elementNameIsEmpty: boolean = elementName === undefined || elementName === '';
```
Comment (robpal94) [inline src/modules/elementname/elementname.service.ts]: See above, parentheses :)

File: src/modules/elementname/elementname.service.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+    if (elementNameIsEmpty) {
+      return '';
+    }
+
+    return '"' + elementName + '"';
```
Comment (robpal94) [inline src/modules/elementname/elementname.service.ts]: Maybe we can use a template string here, such as:
```
return `"${elementName}"`;
```

File: src/modules/processdef-detail/processdef-detail.ts [hunk 25]
```diff
@@ -106,7 +106,11 @@ export class ProcessDefDetail {
     ];
 
     this._eventAggregator.publish(environment.events.navBar.showTools, this.process);
-    this._eventAggregator.publish(environment.events.statusBar.showXMLButton);
+    this._eventAggregator.publish(environment.events.statusBar.showDiagramViewButtons);
+  }
+
+  public determineActivationStrategy(): string {
+    return 'replace';
   }
```
Comment (robpal94) [inline src/modules/processdef-detail/processdef-detail.ts]: Purpose of this method?

File: src/contracts/bpmnmodeler/ICanvas.ts (hunk 9 above)
Comment (SteffenKn) [reply src/contracts/bpmnmodeler/ICanvas.ts]: Imo adding comments to every function makes sense.
For example what comment would you add to `getRootElement(): IShape;` 

I think something like:
```
/*
 * Gets the root element of the canvas.
 *
 * @returns The root element.
 */
getRootElement(): IShape;
```
would not be helpful to anyone

File: src/environment.ts [hunk 26]
```diff
@@ -23,8 +23,8 @@ export default {
     xmlChanged: 'xmlChanged',
     refreshProcessDefs: 'processdefs:refresh',
     statusBar: {
-      showXMLButton: 'statusbar:xmlbutton:show',
-      hideXMLButton: 'statusbar:xmlbutton:hide',
+      showDiagramViewButtons: 'statusbar:diagramviewbuttons:show',
+      hideDiagramViewButtons: 'statusbar:diagramviewbuttons:hide',
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/environment.ts]: That would not fit to the style of the other events.
One could think about refactoring the events, but not in this PR.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts (hunk 8 above)
Comment (SteffenKn) [reply src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Is this really helpful? Then one could also argue that `theDiagramsAreTheSame` is more correct.

File: src/contracts/bpmnmodeler/IEventFunction.ts (hunk 10 above)
Comment (SteffenKn) [reply src/contracts/bpmnmodeler/IEventFunction.ts]: Done

File: src/contracts/diffview/IDiffChanges.ts (hunk 12 above)
Comment (SteffenKn) [reply src/contracts/diffview/IDiffChanges.ts]: Done

File: src/contracts/diffview/IDiffChangeListData.ts (hunk 13 above)
Comment (SteffenKn) [reply src/contracts/diffview/IDiffChangeListData.ts]: What should the comment contain?
I do not know anything useful I could comment here.

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 27]
```diff
@@ -0,0 +1,385 @@
...
+    await this._importXml(coloredXml, this._lowerViewer);
+  }
+
+  private _importXml(xml: string, viewer: IBpmnModeler): Promise <void> {
+    const xmlIsNotLoaded: boolean = (xml === undefined || xml === null);
+
+    if (xmlIsNotLoaded) {
+      const notificationMessage: string = 'The xml could not be loaded. Please try to reopen the Diff View or reload the Detail View.';
+      this._notificationService.showNotification(NotificationType.ERROR, notificationMessage);
+      return;
```
Comment (alexanderkasten) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: newline before return

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 28]
```diff
@@ -0,0 +1,385 @@
...
+    }
+
+    return new Promise((resolve: Function, reject: Function): void => {
+      viewer.importXML(xml, (importXmlError: Error) => {
+        if (importXmlError) {
+          reject(importXmlError);
+          return;
```
Comment (alexanderkasten) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: same like above

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 29]
```diff
@@ -0,0 +1,385 @@
...
+      });
+    });
+  }
+
+  private _getXmlFromModeler(): Promise<string> {
+    return new Promise((resolve: Function, reject: Function): void =>  {
+      this._diffModeler.saveXML({}, async(saveXmlError: Error, xml: string) => {
+        if (saveXmlError) {
+          reject(saveXmlError);
+          return;
```
Comment (alexanderkasten) [inline src/modules/bpmn-diff-view/bpmn-diff-view.ts]: same like above

File: src/modules/bpmn-io/bpmn-io.html [hunk 30]
```diff
@@ -37,10 +38,11 @@
         </div>
       </div>
       <bpmn-xml-view if.bind="initialLoadingFinished" show.bind="showXMLView" xml.bind="xml"></bpmn-xml-view>
+      <bpmn-diff-view if.bind="showDiffView" changes.bind="xmlChanges" savedxml.bind="savedXml" xml.bind="xml"></bpmn-diff-view>
```
Comment (alexanderkasten) [inline src/modules/bpmn-io/bpmn-io.html]: also rename savedxml to savedXml here

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts (hunk 8 above)
Comment (robpal94) [reply src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Whats with `diagramIsUnchanged` ?  🙂

File: src/modules/elementname/elementname.service.ts [hunk 31]
```diff
@@ -0,0 +1,19 @@
+export class ElementNameService {
```
Comment (alexanderkasten) [inline src/modules/elementname/elementname.service.ts]: newline after that

File: src/modules/bpmn-diff-view/bpmn-diff-view.ts [hunk 32]
```diff
@@ -0,0 +1,385 @@
...
+  NotificationType} from '../../contracts/index';
+import environment from '../../environment';
+import {ElementNameService} from '../elementname/elementname.service';
+import { NotificationService } from '../notification/notification.service';
+
+@inject('NotificationService', EventAggregator)
+export class BpmnDiffView {
+
+  @bindable() public xml: string;
+  @bindable({ changeHandler: 'savedXmlChanged' }) public savedxml: string;
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/modules/bpmn-diff-view/bpmn-diff-view.ts]: Camel Case does not work with aurelia binding. I could only use kebab case.

File: src/modules/bpmn-io/bpmn-io.html (hunk 30 above)
Comment (SteffenKn) [reply src/modules/bpmn-io/bpmn-io.html]: Camel Case does not work with aurelia binding. I could only use kebab case.

File: src/modules/elementname/elementname.service.ts (hunk 31 above)
Comment (SteffenKn) [reply src/modules/elementname/elementname.service.ts]: Done

**Pull request merged**